proposals/2021/07-22_log-streaming.md

```diff
@@ -0,0 +1,306 @@
+5. once the service/step is complete, publish remaining logs from the buffer
+
+The end-behavior produced by this method is the logs appear in delayed chunks.
+
+**Please briefly answer the following questions:**
+
+1. Why is this required?
+
+<!-- Answer here -->
+
+* provide compatible functionality with existing CI solutions
+* improve user experience when viewing logs for a service/step
+* improve ability to troubleshoot pipelines by seeing what parts take longer
+
+2. If this is a redesign or refactor, what issues exist in the current implementation?
+
+<!-- Answer here -->
+
+The current implementation leaves a lot to be desired for user experience.
+
+With the logs appearing in delayed chunks, it's difficult to troubleshoot pipelines.
+
+This behavior can make processes appear to be "stuck" or "hung" when inspecting the logs.
+
+This can make it almost impossible to determine if something is running or how long it takes to run.
+
+3. Are there any other workarounds, and if so, what are the drawbacks?
+
+<!-- Answer here -->
+
+N/A
+
+4. Are there any related issues? Please provide them below if any exist.
+
+<!-- Answer here -->
+
+https://github.com/go-vela/community/issues/156
+
+## Design
+
+<!--
+This section is intended to explain the solution design for the proposal.
+
+NOTE: If there are no current plans for a solution, please leave this section blank.
+-->
+
+**Please describe your solution to the proposal. This includes, but is not limited to:**
+
+* new/updated endpoints or url paths
+* new/updated configuration variables (environment, flags, files, etc.)
+* performance and user experience tradeoffs
+* security concerns or assumptions
+* examples or (pseudo) code snippets
+
+<!-- Answer here -->
+
+### Pipeline
+
+To resolve the undesired behavior, we need to create options that can fix it.
+
+However, to craft those options, we need a pipeline that can show the behavior.
+
+We'll use the below pipeline to demonstrate this behavior:
+
+```yaml
+version: "1"
+
+steps:
+  - name: logs
+    image: alpine:latest
+    pull: not_present
+    commands:
+      - sleep 1
+      - echo "hello one"
+      - sleep 2
+      - echo "hello two"
+      - sleep 3
+      - echo "hello three"
+      - sleep 4
+      - echo "hello four"
+      - sleep 5
+      - echo "hello five"
+```
+
+> DISCLAIMER:
+>
+> The options below **only** cover updating code for the **backend** components for log streaming.
+>
+> This means the proposal **does not** cover any updates to the [go-vela/ui](https://github.com/go-vela/ui) codebase.
+>
+> The reason for this is no UI changes are required to produce a streaming effect for logs.
+>
+> This is due to the UI already polling ~ every `5s` in its current state.
+>
+> However, this also means we leave room for improvement in regards to the user experience.
+>
+> This experience could be augmented by making changes to the UI (and server?) along with the below options.
```

Comment:
Do you envision another proposal eventually being submitted for the UI improvements?

Reply:
Yes, I think that could be something we could write up 👍  

FWIW:

As a short-term optimization, we could look to decrease the interval that the UI polls the server?

IIRC, it is currently at ~ `5s` so we could decrease that to something smaller (`2s`? `3s`?).

I'll admit that I'm unsure what the consequences and drawbacks of that would be though.

Reply:
How smart is the current polling setup on the UI? I fear the amount of bandwidth generated by further reducing the poll timeframe.

Reply:
I'm not sure I can accurately speak to that... @wass3r @davidvader thoughts?

I agree with your fears on bandwidth consumption produced from reducing the UI polling 👍 

However, we could certainly explore it if there is enough interest in improving the experience.

Reply:
The current polling is half dumb half smart. (hear me out)

It only polls the logs for steps that are being "viewed" and are not "complete" 
(expanded log trays and not success/failure) as seen [in this code](https://github.com/go-vela/ui/blob/master/src/elm/Main.elm#L2094-L2115)

Then every five seconds it "refreshes the page" and depending on resource, data, status, etc, it "smart polls" from the api. (IE, doesnt refresh a completed step) [this code](https://github.com/go-vela/ui/blob/master/src/elm/Main.elm#L1955-L1961)

This entire polling process is what causes "simulation" and a jarring experience.

**Any type of API polling is going to be a bad experience in my opinion.** 

**For example** step status is updated faster than logs can be produced, so often times we need to continue polling step logs well after the step has already completed. This means we cant use "smart polling" without some type of backoff, which would still be simulated streaming. 

IMO the best solution is always going to be communication from the server TO the UI.
Whether its websockets, httpX, SSE, whatever.

In the interim, we could iterate on the logic in the UI to improve the experience.
- decrease polling interval for logs only
- relax the loading logic on tab switch
- relax the refresh logic on "complete" steps


Im not sure what the best solution is in regards of "effort vs value gained".